packages/backend/src/document.rs

```diff
@@ -95,24 +102,70 @@ pub async fn save_snapshot(state: AppState, data: RefContent) -> Result<(), AppE
     Ok(())
 }
 
-/// Gets an Automerge document ID for the document ref.
 pub async fn doc_id(state: AppState, ref_id: Uuid) -> Result<String, AppError> {
-    let automerge_io = &state.automerge_io;
-    let ack = automerge_io.emit_with_ack::<Vec<Option<String>>>("get_doc", ref_id).unwrap();
-    let mut response = ack.await?;
-
-    let maybe_doc_id = response.data.pop().flatten();
-    if let Some(doc_id) = maybe_doc_id {
-        // If an Automerge doc handle for this ref already exists, return it.
-        Ok(doc_id)
-    } else {
-        // Otherwise, fetch the content from the database and create a new
-        // Automerge doc handle.
-        let content = head_snapshot(state.clone(), ref_id).await?;
-        let data = RefContent { ref_id, content };
-        let ack = automerge_io.emit_with_ack::<Vec<String>>("create_doc", data).unwrap();
-        let response = ack.await?;
-        Ok(response.data[0].to_string())
+    let query = sqlx::query!(
+        "
+        SELECT doc_id FROM refs
+        WHERE id = $1
+        ",
+        ref_id
+    );
+
+    let doc_id =
+        query
+            .fetch_one(&state.db)
+            .await?
+            .doc_id
+            .ok_or(AppError::AutomergeServer(format!(
+                "failed to find doc_id for ref_id '{}'",
+                ref_id
+            )))?;
+
+    start_listening_automerge_doc(&state.automerge_io, ref_id, doc_id.clone()).await?;
+
+    Ok(doc_id)
+}
+
+async fn start_listening_automerge_doc(
+    automerge_io: &SocketIo,
+    ref_id: Uuid,
+    doc_id: String,
+) -> Result<(), AppError> {
+    let ack = automerge_io
+        .emit_with_ack::<Vec<Result<(), String>>>("start_listening", [ref_id.to_string(), doc_id])
+        .map_err(|e| {
+            AppError::AutomergeServer(format!("Failed to call start_listening from backend {}", e))
+        })?;
+
+    let response_array = ack.await?.data;
+    let response = response_array
+        .into_iter()
+        .next()
+        .ok_or_else(|| AppError::AutomergeServer("Empty ack response".to_string()))?;
+
+    match response {
+        Ok(_) => Ok(()),
+        Err(err) => Err(AppError::AutomergeServer(err)),
+    }
+}
+
+async fn create_automerge_doc(automerge_io: &SocketIo, content: Value) -> Result<String, AppError> {
```

Comment:
moved the code for calling SocketIo to a function so that my eyes would bleed less